sys/usb/usbus/cdc/acm/cdc_acm.c

```diff
@@ -0,0 +1,320 @@
+
+static void _handle_flush(event_t *ev);
+
+static const usbus_handler_driver_t cdc_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .control_handler = _control_handler,
+    .transfer_handler = _transfer_handler,
+};
+
+static size_t _gen_full_acm_descriptor(usbus_t *usbus, void *arg);
+
+/* Descriptors */
+static const usbus_hdr_gen_funcs_t _cdcacm_descriptor = {
+    .get_header = _gen_full_acm_descriptor,
+    .len = {
+        .fixed_len = sizeof(usb_desc_cdc_t) +
+                     sizeof(usb_desc_acm_t) +
+                     sizeof(usb_desc_union_t) +
+                     sizeof(usb_desc_call_mngt_t),
+    },
+    .len_type = USBUS_HDR_LEN_FIXED,
+};
+
+static size_t _gen_mngt_descriptor(usbus_t *usbus, usbus_cdcacm_device_t *cdcacm)
+{
+    usb_desc_call_mngt_t mngt;
+    /* functional call management descriptor */
+    mngt.length = sizeof(usb_desc_call_mngt_t);
+    mngt.type = USB_TYPE_DESCRIPTOR_CDC;
+    mngt.subtype = 0x1;
+    mngt.capabalities = 0;
+    mngt.data_if = cdcacm->iface_data.idx;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&mngt, sizeof(mngt));
+    return sizeof(usb_desc_call_mngt_t);
+}
+
+static size_t _gen_union_descriptor(usbus_t *usbus,
+                                    usbus_cdcacm_device_t *cdcacm)
+{
+    usb_desc_union_t uni;
+    /* functional union descriptor */
+    uni.length = sizeof(usb_desc_union_t);
+    uni.type = USB_TYPE_DESCRIPTOR_CDC;
+    uni.subtype = 0x6;
+    uni.master_if = cdcacm->iface_ctrl.idx;
+    uni.slave_if = cdcacm->iface_data.idx;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&uni, sizeof(uni));
+    return sizeof(usb_desc_union_t);
+}
+
+static size_t _gen_acm_descriptor(usbus_t *usbus)
+{
+    usb_desc_acm_t acm;
+    /* functional cdc acm descriptor */
+    acm.length = sizeof(usb_desc_acm_t);
+    acm.type = USB_TYPE_DESCRIPTOR_CDC;
+    acm.subtype = 0x02;
+    acm.capabalities = 0x00;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&acm, sizeof(acm));
+    return sizeof(usb_desc_acm_t);
+}
+
+static size_t _gen_cdc_descriptor(usbus_t *usbus)
+{
+    usb_desc_cdc_t cdc;
+    /* functional cdc descriptor */
+    cdc.length = sizeof(usb_desc_cdc_t);
+    cdc.bcd_cdc = 0x0120;
+    cdc.type = USB_TYPE_DESCRIPTOR_CDC;
+    cdc.subtype = 0x00;
+    usbus_control_slicer_put_bytes(usbus, (uint8_t*)&cdc, sizeof(cdc));
+    return sizeof(usb_desc_cdc_t);
+}
+
+static size_t _gen_full_acm_descriptor(usbus_t *usbus, void *arg)
+{
+    usbus_cdcacm_device_t *cdcacm = (usbus_cdcacm_device_t*)arg;
+    size_t total_len = 0;
+    total_len += _gen_cdc_descriptor(usbus);
+    total_len += _gen_acm_descriptor(usbus);
+    total_len += _gen_union_descriptor(usbus, cdcacm);
+    total_len += _gen_mngt_descriptor(usbus, cdcacm);
+    return total_len;
+}
+
+/* Submit (ACM interface in) */
+size_t usbus_cdc_acm_submit(usbus_cdcacm_device_t *cdcacm, const uint8_t *buf, size_t len)
+{
+    return tsrb_add(&cdcacm->tsrb, buf, len);
```

Comment:
Note: I'm going to have to modify this to a `tsrb_add_one()` wrapped in a `while(buf)`. `tsrb_add_one()` is not blocking and then it can break as soon as the tsrb is full. This will prevent a hang when the `tsrb` is full. 

The downside is that it will only buffer the first n chars of data while I would prefer to buffer the last n chars until the buffer is full. But that can be done in a follow up.

Reply:
> Note: I'm going to have to modify this to a `tsrb_add_one()` wrapped in a `while(buf)`. `tsrb_add_one()` is not blocking and then it can break as soon as the tsrb is full. This will prevent a hang when the `tsrb` is full.

Never mind this, I had some misunderstanding about the working of `tsrb_add()`